common/src/api/external/mod.rs

```diff
@@ -839,26 +850,43 @@ impl From<u32> for Generation {
 }
 
 impl TryFrom<i64> for Generation {
-    type Error = anyhow::Error;
+    type Error = GenerationNegativeError;
 
     fn try_from(value: i64) -> Result<Self, Self::Error> {
         Ok(Generation(
-            u64::try_from(value)
-                .map_err(|_| anyhow!("negative generation number"))?,
+            u64::try_from(value).map_err(|_| GenerationNegativeError(()))?,
         ))
     }
 }
 
 impl TryFrom<u64> for Generation {
-    type Error = anyhow::Error;
+    type Error = GenerationOverflowError;
 
     fn try_from(value: u64) -> Result<Self, Self::Error> {
-        i64::try_from(value)
-            .map_err(|_| anyhow!("generation number too large"))?;
+        i64::try_from(value).map_err(|_| GenerationOverflowError(()))?;
         Ok(Generation(value))
     }
 }
 
+impl FromStr for Generation {
+    type Err = std::num::ParseIntError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        // Try to parse `s` as both an i64 and u64, returning the error from
+        // either.
+        let _ = i64::from_str(s)?;
+        Ok(Generation(u64::from_str(s)?))
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+#[error("negative generation number")]
+pub struct GenerationNegativeError(());
+
+#[derive(Debug, thiserror::Error)]
+#[error("generation number too large")]
```

Comment:
this error, on the other hand, _could_ interpolate the value of `Generation::MAX` into its message when it's formatted. if you want to.